### ISSUE TITLE
[Installation][non-reproducible]: Op Flash Attention

### ISSUE THREAD

**Issue by antferdom** [pytorch-labs/tritonbench#17]
The current project repository **assumes** existing **submodules** directory for all the optional dependencies. The Python installation script executes `checkout_submodules` but that’s again only relevant if submodule dir is populated accordingly (`git submodule add`). The following represents the expected `.gitmodules`:
```ruby
[submodule "submodules/FBGEMM"]
	path = submodules/FBGEMM
	url = https://github.com/pytorch/fbgemm
[submodule "submodules/flash-attention"]
	path = submodules/flash-attention
	url = https://github.com/Dao-AILab/flash-attention.git
[submodule "submodules/ThunderKittens"]
	path = submodules/ThunderKittens
	url = https://github.com/HazyResearch/ThunderKittens.git
[submodule "submodules/cutlass-kernels"]
	path = submodules/cutlass-kernels
	url = https://github.com/ColfaxResearch/cutlass-kernels.git
[submodule "submodules/generative-recommenders"]
	path = submodules/generative-recommenders
	url = https://github.com/facebookresearch/generative-recommenders.git
[submodule "submodules/kernels"]
	path = submodules/kernels
	url = https://github.com/triton-lang/kernels.git
[submodule "submodules/cutlass"]
	path = submodules/cutlass
	url = https://github.com/NVIDIA/cutlass.git
```
Then we execute `git submodule update --init --recursive`.

# Dockerfile
Two final steps of the Docker image building are commented because they reference bash installation scripts from a non-existing `.ci` directory, presumably available  in the Meta's internal repo. The commented Dockerfile:
```Dockerfile
# Tritonbench library build and test require libcuda.so.1
# which is from NVIDIA driver
RUN sudo apt update && sudo apt-get install -y libnvidia-compute-550 patchelf

# Install Tritonbench
#RUN cd /workspace/tritonbench && \
#    bash .ci/tritonbench/install.sh


# Test Tritonbench
#RUN cd /workspace/tritonbench && \
#    bash .ci/tritonbench/test-install.sh

# Remove NVIDIA driver library - they are supposed to be mapped at runtime
RUN sudo apt-get purge -y libnvidia-compute-550

# Clone the pytorch env as triton-main env, then compile triton main from source
RUN cd /workspace/tritonbench && \
    BASE_CONDA_ENV=${CONDA_ENV} CONDA_ENV=${CONDA_ENV_TRITON_MAIN}
    # bash .ci/tritonbench/install-triton-main.sh

```
Instead, we execute the container in interactive mode (`docker exec -it`), and based on the comment of that script, **build Triton from source**.

@xuzhao9

**Comment by xuzhao9** [comment]
Thanks for reporting the issue. Sorry that there are some sync issues going on and we are working on fixing it.

**Comment by antferdom** [comment]
Is there anything I can do to help?

**Comment by xuzhao9** [comment]
@antferdom Can you help verify the problem is fixed by https://github.com/pytorch-labs/tritonbench/commit/748a883273a90b5b6499e8a2e927bb40ae0084de ?

**Comment by antferdom** [comment]
@xuzhao9 sure, I will try it out tomorrow

**Comment by antferdom** [comment]
@xuzhao9 

- Commit cebd61df0355949580c1476d9bd84a42fab1108e modified several `run.py` imports, thus now or we install [pytorch.benchmark](https://github.com/pytorch/benchmark) as additional dependency, or it won’t run (see `from tritonbench.utils.env_utils import AVAILABLE_PRECISIONS`)
- Correct Docker image building installing all the optional dependencies in the git submodules.
- Tritonbench is not installed as a Python module in the current commit.
- Xformers is not built from source. This is crucial if we want to leverage the xformers backend with the [Torch custom op wrapper logic for FlasthAttention v3](https://github.com/facebookresearch/xformers/blob/a97a1e0cb4337b0a8a8ee1670618a1f15fb84f54/xformers/ops/fmha/flash3.py#L101)
```bash
memory_efficient_attention.fa2F@v2.5.6-pt:         available
memory_efficient_attention.fa2B@v2.5.6-pt:         available
memory_efficient_attention.fa3F@0.0.0:             available
memory_efficient_attention.fa3B@0.0.0:             available
```

- Building Triton from source instal-triton-main.sh leads to TMA errors (e.g `run.py –op flash_attention –mode fwd`). Removing the installation and installing it with the latest nightly release, solves the issue. Nevertheless, the execution of flash_attention op halts indefinitely after the first iteration, without printing any error. Although Torch compilation could be a potential reason for this long execution time, even after 1h, still on the first benchmarking iteration.

**op:** `flash_attention`

```bash
(pytorch) runner@compiler-study-hopper:/workspace/tritonbench$ python run.py --op flash_attention --mode fwd
TMA benchmarks will be running without grid constant TMA descriptor.
 11%|████████████████▉                                                                                                                                       | 1/9 [00:10<01:25, 10.67s/it]
```

**Comment by xuzhao9** [comment]
> * Commit [cebd61d](https://github.com/pytorch-labs/tritonbench/commit/cebd61df0355949580c1476d9bd84a42fab1108e) modified several `run.py` imports, thus now or we install [pytorch.benchmark](https://github.com/pytorch/benchmark) as additional dependency, or it won’t run (see `from tritonbench.utils.env_utils import AVAILABLE_PRECISIONS`)
> * Correct Docker image building installing all the optional dependencies in the git submodules.
> * Tritonbench is not installed as a Python module in the current commit.

The above three errors should be fixed by https://github.com/pytorch-labs/tritonbench/commit/7e60e23471ce98ea1d947b7df27fd9290fa660dc.

> * Xformers is not built from source. This is crucial if we want to leverage the xformers backend with the [Torch custom op wrapper logic for FlasthAttention v3](https://github.com/facebookresearch/xformers/blob/a97a1e0cb4337b0a8a8ee1670618a1f15fb84f54/xformers/ops/fmha/flash3.py#L101)
> 
> ```shell
> memory_efficient_attention.fa2F@v2.5.6-pt:         available
> memory_efficient_attention.fa2B@v2.5.6-pt:         available
> memory_efficient_attention.fa3F@0.0.0:             available
> memory_efficient_attention.fa3B@0.0.0:             available
> ```

Create https://github.com/pytorch-labs/tritonbench/issues/20 to track the progress.

> * Building Triton from source instal-triton-main.sh leads to TMA errors (e.g `run.py –op flash_attention –mode fwd`). Removing the installation and installing it with the latest nightly release, solves the issue. Nevertheless, the execution of flash_attention op halts indefinitely after the first iteration, without printing any error. Although Torch compilation could be a potential reason for this long execution time, even after 1h, still on the first benchmarking iteration.
> 
> **op:** `flash_attention`
> 
> ```shell
> (pytorch) runner@compiler-study-hopper:/workspace/tritonbench$ python run.py --op flash_attention --mode fwd
> TMA benchmarks will be running without grid constant TMA descriptor.
>  11%|████████████████▉                                                                                                                                       | 1/9 [00:10<01:25, 10.67s/it]
> ```
> 

In the docker image, we are using different conda environments to manage the Triton versions:
The `pytorch` env uses the triton version built-in with the latest pytorch nightly release.
The `triton-main` env uses the Triton main branch.
Can you please try running the command with the `triton-main` env? If the problem still exists, please create another issue to track.


> * **ThunderKittens** installation wrong shared object path:
> 
> ```python
> In [2]: from tritonbench.utils.loader import load_library
> 
> In [3]: load_library("tk/tk_attn_h100_fwd.so")
> ---------------------------------------------------------------------------
> OSError                                   Traceback (most recent call last)
> Cell In[3], line 1
> ----> 1 load_library("tk/tk_attn_h100_fwd.so")
> 
> File /workspace/tritonbench/tritonbench/utils/loader.py:9, in load_library(library_path)
>       7 prefix, _delimiter, so_file = library_path.partition("/")
>       8 so_full_path = REPO_PATH.joinpath(prefix, ".data", so_file).resolve()
> ----> 9 torch.ops.load_library(str(so_full_path))
> 
> File ~/miniconda3/envs/pytorch/lib/python3.11/site-packages/torch/_ops.py:1357, in _Ops.load_library(self, path)
>    1352 path = _utils_internal.resolve_library_path(path)
>    1353 with dl_open_guard():
>    1354     # Import the shared library into the process, thus running its
>    1355     # static (global) initialization code in order to register custom
>    1356     # operators with the JIT.
> -> 1357     ctypes.CDLL(path)
>    1358 self.loaded_libraries.add(path)
> 
> File ~/miniconda3/envs/pytorch/lib/python3.11/ctypes/__init__.py:376, in CDLL.__init__(self, name, mode, handle, use_errno, use_last_error, winmode)
>     373 self._FuncPtr = _FuncPtr
>     375 if handle is None:
> --> 376     self._handle = _dlopen(self._name, mode)
>     377 else:
>     378     self._handle = handle
> 
> OSError: /workspace/tritonbench/tritonbench/tk/.data/tk_attn_h100_fwd.so: cannot open shared object file: No such file or directory
> ```
> 
> correct path -> `/workspace/tritonbench/utils/tk/.data/tk_attn_h100_fwd.so`, there seems to be an additional **tritonbench**. Using this full path instead of the utility function for loading so:
> 
> ```python
> In [1]: import torch
> 
> In [2]: torch.ops.load_library("/workspace/tritonbench/utils/tk/.data/tk_attn_h100_fwd.so")
> 
> In [3]:  tk_fwd = torch.ops.tk
> ```

I am working on this, will test and have a PR soon.

**Comment by xuzhao9** [comment]
For the flash-attention operator there are two problems:

1. triton_fused_attention kernel does not work for triton-main: we do not plan to fix that soon (soon means maybe in one or two weeks), as there might be many other kernels that don't work for triton-main at this time, too. We plan to first build the CI and systematically find out how many of the existing kernels that don't work for triton-main. Especially, we need PT2-team to get involved to if there is incompatibility issues between Torchinductor and Triton-main.

3. flash_attention on pytorch built-in Triton should work ootb and is a problem if any of them hang, I will take a look at this.

**Comment by antferdom** [comment]
1. **triton_fused_attention kernel does not work for triton-main:** Understood, sounds like a good formal procedure to isolate non working kernels with `triton-main`
2. It works, but after the first iteration is completed (reported by `tqdm`) the execution halts indefinitely (let it running in the mentioned environment for around 1h). GPU utilisation is at maximum capacity, but there isn't any memory alteration. I will try to isolate and go one by one attention implementation in a standalone script without the operator registry logic of Tritonbench-> [test_attn](https://github.com/ai-compiler-study/test_attn)

**Comment by xuzhao9** [comment]
For 2), you could use `--only` to specify a single attention impl, this should be easier to bisect/isolate.

**Comment by antferdom** [comment]
True, I will use the `--only` to isolate one implementation step by step.

**Comment by antferdom** [comment]
Using the latest Tritonbench Docker image, I was finally able to fully run the `flash_attention` operator:
```bash
/workspace/tritonbench$ python run.py --op flash_attention --mode fwd --precision bf16
TMA benchmarks will be running without grid constant TMA descriptor.
100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 9/9 [01:05<00:00,  7.25s/it]
  (Batch, Heads, SeqLen, Dhead)    sdpa-latency    aten-latency    flash_v2-latency    flash_v3-latency    triton_tutorial_flash_v2-latency    triton_tutorial_flash_v2_tma-latency    flex_attention-latency
-------------------------------  --------------  --------------  ------------------  ------------------  ----------------------------------  --------------------------------------  ------------------------
              (32, 32, 512, 64)        0.262816        2.88429             0.253344            0.212928                            0.204864
             (16, 32, 1024, 64)        0.483936        5.64954             0.461056            0.360672                            0.371744
              (8, 32, 2048, 64)        0.928736       13.2899              0.900224            0.621792                            0.685088
              (4, 32, 4096, 64)        1.81328        23.6284              1.74131             1.18486                             1.38013
              (2, 32, 8192, 64)        3.59718        45.2292              3.44819             2.19107                             2.62384
             (1, 32, 16384, 64)        7.15782                             6.85792             4.28938                             5.21392                                                           5.20269
               (4, 32, 19, 128)        0.009248        0.032416            0.01328             0.012032                            0.00832                                                           0.099776
                (4, 32, 1, 128)        0.009056        0.025888            0.013888            0.01152                             0.007904                                                          0.098656
              (4, 32, 511, 128)        0.066976        0.54688             0.065312            0.048288                            0.060288                                                          0.082976
```

Nevertheless, many attention implementations are registered as `enabled=False`:

- xformers
- xformers_splitk
- colfax_cutlass
- cudnn
- thunderkittens

**Comment by xuzhao9** [comment]
https://github.com/pytorch-labs/tritonbench/pull/58 fixes as many the impls as we can.
AFAIK only TK is broken because it won't accept the input shapes we provide (it requires SEQ_LEN must be multiplies of 192).

**Comment by antferdom** [comment]
Thanks for the update @xuzhao9, can we close this issue once #58 is merged? Can you enable that hook for this issue?

**Comment by xuzhao9** [comment]
> Thanks for the update @xuzhao9, can we close this issue once #58 is merged? Can you enable that hook for this issue?

Yes, we can close this issue now.
